### PR TITLE
Calculate metrics deltas in metrics_collector.js

### LINKED ISSUE #966: Calculate counter deltas client side for the admin dashboard
The admin dashboard makes calls to `/admin/metrics` every second to get delta values for a subset of counters.  It would simplify the metrics interface to not need to provide this and only provide `/admin/metrics.json` (the full list of metrics without any deltas).  The admin dashboard could store the previous value for counters and calculate diffs on the client side.

The deltas from `/admin/metrics` are not on a strict 1 second window, they refresh only when queried (up to a maximum of once per second).  Therefore, I don't think moving the delta calculation to the client will result in any loss of accuracy since the windowing was already driven by the rate at which the client queried.  

The result is that the server will no longer need to make any assumptions about the rate at which clients are querying.

### PR DESCRIPTION
**Pull request by esbie**
Problem:
We'd like to drop support of delta values in `admin/metrics.json`.

Solution:
Calculate deltas client-side instead.
Fixes #966

This is a very surgical change in the interest of #1040 
Going forward I'd like to add a metrics endpoint that better reflects the MetricsTree structure. This would allow the frontend to simplify `routers.js` and any components currently using `query.js`

### CODE REVIEW

File: admin/src/main/resources/io/buoyant/admin/js/src/metrics_collector.js [hunk 1]
```diff
@@ -38,46 +27,56 @@ define(['jQuery'], function($) {
       _.remove(listeners, function(l) { return l.handler === handler; });
     }
 
-    return function(initialMetrics) {
-      var defaultMetrics = _.keys(initialMetrics);
+    function generateDeltaPayload(generalData, defaultMetrics, prevMetrics) {
+      var metrics = _(listeners)
+        .map(function(listener){ return listener.metrics(defaultMetrics); })
+        .flatten()
+        .uniq()
```
Comment (rmars) [inline admin/src/main/resources/io/buoyant/admin/js/src/metrics_collector.js]: a `.compact()` here might be nice to get rid of the `""` and `undefined`s

File: admin/src/main/resources/io/buoyant/admin/js/src/metrics_collector.js [hunk 2]
```diff
@@ -38,46 +27,56 @@ define(['jQuery'], function($) {
       _.remove(listeners, function(l) { return l.handler === handler; });
```
Comment (rmars) [inline admin/src/main/resources/io/buoyant/admin/js/src/metrics_collector.js]: I would also maybe update the description of `specific: {}` in the comment

**Comment by adleong** [comment]
Created https://github.com/linkerd/linkerd/issues/1055 to track.

**Comment by esbie** [comment]
Oh interesting! The `?m=` endpoint has a `delta` field for all stat types, but the field behaves differently based on stat type. 
The client doesn't receive any stat type information, so it'd be difficult to imitate this behavior (and also, the behavior is kind of crazy IMO so I don't want to recreate it). I think instead I'll update the RouterSummary and RouterClient components to be gauge-aware for the gauges they use. Then as part of #1055 we can send stat type in the api payload.

File: admin/src/main/resources/io/buoyant/admin/js/src/router_client.js [hunk 3]
```diff
@@ -68,7 +68,7 @@ define([
     function getMetricDefinitions(routerName, clientName) {
       return _.map([
           {suffix: "requests", label: "Requests"},
-          {suffix: "connections", label: "Connections"},
+          {suffix: "connections", label: "Connections", isGauge: true},
           {suffix: "success", label: "Successes"},
           {suffix: "failures", label: "Failures"},
           {suffix: "loadbalancer/size", label: "Load balancer pool size"},
```
Comment (adleong) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_client.js]: I believe these load balancer metrics are gauges also

Comment (esbie) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_client.js]: good call

**Comment by esbie** [comment]
Hm ok, show me your setup when you get a chance; connections look fine in my tests.

screen shot 2017-02-09 at 10 07 53 am

**Comment by adleong** [comment]
Oh, it seems to be working now!  👍

**Comment by esbie** [comment]
yay thank you for checking!

**Pull request merged**